### PR TITLE
Add UI devhub login and upload tests

### PR DESCRIPTION
**Pull request by jrbenny35**
This pull request will add 3 new tests specifically testing the devhub, as well as the infamous, Devhub Upload. 🔥

1) ```test_devhub_loads_addons```: This test will check that the number of addons shown when a user logs into their devhub account, matches the number that the ```api``` returns.
2) ```test_devhub_addon_edit_link_works```: This will check the edit link for editing an addon loads the edit page.
3) ```test_devhub_addon_upload```: This will create an addon by uploading a ```webextension.xpi``` and filling out the submission form. It will also skip the admin review process.

*** Future plans
I'd like to really test the fill flow of addon upload and installation. As discussed during the work week, we plan to implement a CRON feature for auto approving addons and allow it to be run on the docker image.

### CODE REVIEW

**Comment by jrbenny35** [comment]
@EnTeQuAk looks like there was an upload error. It says ```Received an empty response from the server: Status 500```. If you look at the artifact html report you can see where it says that. Any ideas?

File: tests/ui/conftest.py [hunk 1]
```diff
@@ -19,13 +14,14 @@ def firefox_options(firefox_options):
...
-    firefox_options.add_argument('-headless')
-    firefox_options.log.level = 'trace'
+        "extensions.install.requireBuiltInCerts", False
+    )
+    firefox_options.set_preference("xpinstall.signatures.required", False)
+    firefox_options.set_preference("extensions.webapi.testing", True)
+    firefox_options.set_preference("ui.popup.disable_autohide", True)
+    firefox_options.add_argument("-foreground")
+    firefox_options.add_argument("-headless")
+    firefox_options.log.level = "trace"
```
Comment (EnTeQuAk) [inline tests/ui/conftest.py]: We are using single quotes throughout our whole code, let's not change this needlessly.

File: Makefile-docker [hunk 2]
```diff
@@ -148,13 +148,12 @@ setup-ui-tests:
 	# Create a proper superuser that can be used to access the API
 	python manage.py waffle_switch super-create-accounts on --create
 	python manage.py waffle_switch activate-autograph-signing on --create
-	python manage.py createsuperuser --email=uitest@mozilla.com --username=uitest --noinput --add-to-supercreate-group --save-api-credentials=tests/ui/variables.json --hostname=olympia-frontend.test
+	python manage.py createsuperuser --email=uitest@restmail.net --username=uitest --noinput --add-to-supercreate-group --save-api-credentials=tests/ui/variables.json --hostname=olympia-frontend.test
```
Comment (EnTeQuAk) [inline Makefile-docker]: If we're not making use of the saved variables anymore we should remove that functionality imho.

File: tests/ui/conftest.py [hunk 3]
```diff
@@ -19,13 +14,14 @@ def firefox_options(firefox_options):
 
     """
     firefox_options.set_preference(
-        'extensions.install.requireBuiltInCerts', False)
```
Comment (EnTeQuAk) [inline tests/ui/conftest.py]: All these options still need some documentation, as discussed, can you add that please?

File: tests/ui/conftest.py [hunk 4]
```diff
@@ -47,61 +56,38 @@ def selenium(selenium, request):
     Mobile size: 738x414 (iPhone 7+)
 
     """
+    # Skip mobile test with marker 'desktoponly'
+    if request.node.get_marker("desktoponly") and request.param == (414, 738):
+        pytest.skip("Skipping mobile test")
```
Comment (EnTeQuAk) [inline tests/ui/conftest.py]: That's kinda ugly. Let's move the resolution params into constants and re-use them when needed.

```python
DESKTOP = (1080, 1920)
MOBILE = (414, 738)

if request.node.get_marker('desktoponly') and request.param == MOBILE:
    # ...
```

File: Makefile-docker (hunk 2 above)
Comment (jrbenny35) [reply Makefile-docker]: Remove it from the api?

**Comment by jrbenny35** [comment]
@EnTeQuAk have you gotten a chance to look into this a little?

File: Makefile-docker (hunk 2 above)
Comment (EnTeQuAk) [reply Makefile-docker]: Also, yes.

**Comment by EnTeQuAk** [comment]
When running this locally I get this traceback: https://gist.githubusercontent.com/EnTeQuAk/e7b01a3e9df279bd526f330f5136c775/raw/fdf53d45dd7bb897c728ddc352485349724c7daa/gistfile1.txt

I assume it's related to permissions to the `tmp` and `user-media` folders, especially when looking at the error, `IOError: [Errno 13] Permission denied: '/code/user-media/addons/temp/0cb43ccc15664f9fb7d2574bf7ffc637.xpi'` this shows us that the runserver process (`web` as well as the `worker` container) simply rely on `settings.py` and don't use any temporary storage location located in `/tmp/` which is probably much more open and easier to write to.

Something else on the interwebs https://support.circleci.com/hc/en-us/articles/360003649774-Permission-Denied-When-Creating-Directory-or-Writing-a-File (not necessarily related, there might be more)

The problem is (I think) that we're trying to write to a directory we don't have permissions to. For our travis tests we have something like https://github.com/mozilla/addons-server/blob/master/conftest.py#L104-L106 that uses the `tmpdir` pytest fixture but that won't work for our ui-tests because these settings live in the `web` and `worker` container and they're using `settings.py` afaik. So we'd have to provide our own settings, add a docker-compose override for ui-tests to provide those settings and correctly configure the web and worker containers

You can try just changing the settings.py file in your branch and point these 3 settings to a temporarily created folder in /tmp/ to test my theory if that works… then we can try to figure out a more correct approach forward.

We can't use the pytest fixture or `tmpdir` because the runserver process is outside of pytest. It's being started as part of `docker-compose up -d` in the circle.yml.

So we'll have to create a folder in /tmp/ and point at least `settings.MEDIA_ROOT` (which should be something like `/tmp/<random-foo-bar>/media/`), `settings.TMP_PATH` (which should be something like `/tmp/<random-foo-bar>/tmp/`) to the newly created folder. Also make sure `settings.NETAPP_STORAGE` is set to `settings.TMP_PATH` (the *newly configured value*).

https://github.com/mozilla/addons-server/blob/master/conftest.py#L104-L106 as inspiration but as I noted, we can't reuse that because the runserver process lives outside of pytest.

**Comment by jrbenny35** [comment]
I was experimenting with these settings:
```
MEDIA_ROOT = path('tmp/ui-tests/user-media/')
# MEDIA_URL = 'tmp/ui-tests/user-media/'
TMP_PATH = path('tmp/ui-tests/tmp/')
NETAPP_STORAGE = TMP_PATH
FILESYSTEM_CACHE_ROOT = path('tmp/ui-tests/tmp/cache/')
```
It seems that now I get an error about a cache missing. I tried making it but that doesn't seem to work.

File: scripts/ui-test-hostname-setup.sh [hunk 5]
```diff
@@ -2,7 +2,7 @@
 # This script sets the ip of the addons-frontend image as localhost within the selenium-firefox image.
 # This MUST be run before any user integration tests.
 
-UI_IP="`docker inspect addons-server_addons-frontend_1 | grep "IPAddress" | grep -oE "\b([0-9]{1,3}\.){3}[0-9]{1,3}\b"`"
+UI_IP="`docker inspect addonsserver_addons-frontend_1 | grep "IPAddress" | grep -oE "\b([0-9]{1,3}\.){3}[0-9]{1,3}\b"`"
```
Comment (jrbenny35) [inline scripts/ui-test-hostname-setup.sh]: This is changed because I noticed it was different from my local config, so we now install docker-compose using curl, and a local tester should be able to use this script as well.

File: Makefile-docker (hunk 2 above)
Comment (EnTeQuAk) [reply Makefile-docker]: This seems still todo?

This needs to be removed from the management command, there is nothing in the API actually. So `src/olympia/users/management/commands/createsuperuser.py` is the only place where the variable saving should be removed from.

File: circle.yml [hunk 6]
```diff
@@ -82,7 +82,9 @@ jobs:
           name: Install Docker Compose
           command: |
             set -x
-            pip install docker-compose>=1.18
+            curl -L https://github.com/docker/compose/releases/download/1.19.0/docker-compose-`uname -s`-`uname -m` > ~/docker-compose
+            chmod +x ~/docker-compose
+            sudo mv ~/docker-compose /usr/local/bin/docker-compose
```
Comment (EnTeQuAk) [inline circle.yml]: Why are we doing this compared to installing via pip? Maybe add a comment?

File: circle.yml [hunk 7]
```diff
@@ -99,8 +101,9 @@ jobs:
             docker-compose restart worker
             docker-compose exec web make -f Makefile-docker update_deps
             docker-compose restart web
+            sudo chown -R circleci:circleci .
```
Comment (EnTeQuAk) [inline circle.yml]: Add a comment why we do this.

File: tests/ui/conftest.py (hunk 3 above)
Comment (EnTeQuAk) [reply tests/ui/conftest.py]: Can you either open an issue about that or simply add comments that document these settings one by one?

File: tests/ui/conftest.py [hunk 8]
```diff
@@ -47,61 +60,38 @@ def selenium(selenium, request):
...
-def jwt_issuer(base_url, variables):
-    """JWT Issuer from variables file or env variable named 'JWT_ISSUER'"""
-    try:
-        hostname = urlparse.urlsplit(base_url).hostname
-        return variables['api'][hostname]['jwt_issuer']
-    except KeyError:
-        return os.getenv('JWT_ISSUER')
-
+def base_url(base_url):
+    return base_url
```
Comment (EnTeQuAk) [inline tests/ui/conftest.py]: Do we actually need this if it's just a dummy like that? :thinking:

File: tests/ui/pages/desktop/devhub.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    def addons_list(self):
+        els = self.find_elements(*self._addons_item_locator)
+        return [self.AddonsListItem(self, el) for el in els]
+
+    def upload_addon(self):
+        """Upload an addon via devhub.
+
+        This will use the override validation option.
+        """
+        file_to_upload = 'webextension_no_id.xpi'
```
Comment (EnTeQuAk) [inline tests/ui/pages/desktop/devhub.py]: I think it's worth making this an argument of this method. That'll allow us to simply upload other add-ons in different ui-tests for edge-cases or so.

File: tests/ui/pages/desktop/devhub.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        file_path = get_file(file_to_upload)
+        self.selenium.find_element(*self._submit_addon_btn_locator).click()
+        self.selenium.find_element(*self._continue_sub_btn_locator).click()
+        upload = self.selenium.find_element(*self._upload_addon_locator)
+        upload.send_keys(file_path)
+        self.wait.until(
+            expected.element_to_be_clickable(self._override_validation_locator)
+        )
+        self.selenium.find_element(*self._override_validation_locator).click()
```
Comment (EnTeQuAk) [inline tests/ui/pages/desktop/devhub.py]: Why do you override validation errors? This shouldn't be possible for regular users anyway so the ui-tests should not rely on them. If you're running into errors with the linter, either use a different fixture that don't have them or just fix them?

File: tox.ini [hunk 11]
```diff
@@ -17,51 +17,50 @@ whitelist_externals =
 
 [testenv:es]
 commands =
-    make -f Makefile-docker install_python_test_dependencies
-    pytest -m "es_tests and not needs_locales_compilation and not static_assets" --ignore=tests/ui/ -v src/olympia/{posargs}
+    make -f Makefile-docker update_deps
+    pytest -m "es_tests and not needs_locales_compilation" --ignore=tests/ui/ -v {posargs}
```
Comment (EnTeQuAk) [inline tox.ini]: Please don't change this, looks like a merge gone bad? The same for all the other environments in the `tox.ini`. We don't want to run `update_deps` directly here for performance reasons.

File: tox.ini [hunk 12]
```diff
@@ -17,51 +17,50 @@ whitelist_externals =
...
 commands =
     make -f Makefile-docker update_deps
     make -f Makefile-docker ui-tests
-    pytest --driver Firefox -v tests/ui/ {posargs}
+    pytest --driver Firefox tests/ui/ {posargs}
 
 [testenv:assets]
 commands =
     make -f Makefile-docker update_deps
-    pytest -m "static_assets" --ignore=tests/ui/ -v src/olympia/ {posargs}
```
Comment (EnTeQuAk) [inline tox.ini]: This also should not get removed.

File: circle.yml (hunk 6 above)
Comment (jrbenny35) [reply circle.yml]: This is changed because I noticed it was different from my local config, so we now install docker-compose using curl, and a local tester should be able to use this script as well. 

For the script within ```scripts/ui-hostname-setup.sh``` we had a different config from the usual localenv. Well maybe it was the same, but I noticed on my mac, which has docker compose installed via Pip the names given to the container contained no spaces or ```-```'s.
```addonsserver_addons-frontend_1``` vs ```addons-server_addons-frontend_1```.

File: tests/ui/conftest.py (hunk 8 above)
Comment (jrbenny35) [reply tests/ui/conftest.py]: Nope!

Comment (jrbenny35) [reply tests/ui/conftest.py]: Its so lonely 😨

File: tests/ui/pages/desktop/devhub.py (hunk 10 above)
Comment (jrbenny35) [reply tests/ui/pages/desktop/devhub.py]: I guess I can do this I just didn't feel testing the lint of the extensions was needed.

File: circle.yml (hunk 7 above)
Comment (EnTeQuAk) [reply circle.yml]: This comment only explains what it does, it doesn't explain why we have to change permissions. Also, you're not just changing permissions of `olympia.egg*` but recursively down all folders starting from the current one.

File: tests/ui/conftest.py [hunk 13]
```diff
@@ -24,7 +33,7 @@ def firefox_options(firefox_options):
     firefox_options.set_preference('extensions.webapi.testing', True)
     firefox_options.set_preference('ui.popup.disable_autohide', True)
     firefox_options.add_argument('-foreground')
-    firefox_options.add_argument('-headless')
+    # firefox_options.add_argument('-headless')
```
Comment (EnTeQuAk) [inline tests/ui/conftest.py]: If unused this should be removed instead of commented out.

File: tox.ini [hunk 14]
```diff
@@ -17,51 +17,51 @@ whitelist_externals =
...
     make -f Makefile-docker update_deps
     make -f Makefile-docker ui-tests
-    pytest --driver Firefox tests/ui/ {posargs}
+    pytest --driver Firefox tests/ui/ -k test_devhub_addon_upload {posargs}
 
 [testenv:assets]
 commands =
     make -f Makefile-docker update_deps
-    make -f Makefile-docker update_assets
+    pytest -m "static_assets" --ignore=tests/ui/ -v src/olympia/ {posargs}
```
Comment (EnTeQuAk) [inline tox.ini]: We are running `update_assets` after running `pytest` here on master. Can you add that back please?

File: tox.ini [hunk 15]
```diff
@@ -17,51 +17,51 @@ whitelist_externals =
...
-    pytest -n 2 -m 'not es_tests and not needs_locales_compilation' -v src/olympia/ --ignore src/olympia/addons/ --ignore src/olympia/devhub/ --ignore src/olympia/reviewers/ --ignore src/olympia/amo/ --ignore src/olympia/users/ --ignore src/olympia/accounts/ --ignore src/olympia/lib/crypto --ignore src/olympia/signing {posargs}
+    make -f Makefile-docker install_python_test_dependencies install_node_dependencies
+    pytest -n 2 -m 'not es_tests and not needs_locales_compilation and not static_assets' -v src/olympia/ --ignore src/olympia/addons/ --ignore src/olympia/devhub/ --ignore src/olympia/reviewers/ --ignore src/olympia/amo/ --ignore src/olympia/users/ --ignore src/olympia/accounts/ --ignore src/olympia/lib/crypto --ignore src/olympia/signing {posargs}
 
 [testenv:ui-tests]
 commands =
     make -f Makefile-docker update_deps
     make -f Makefile-docker ui-tests
-    pytest --driver Firefox tests/ui/ {posargs}
+    pytest --driver Firefox tests/ui/ -k test_devhub_addon_upload {posargs}
```
Comment (EnTeQuAk) [inline tox.ini]: Should probably run all tests? :see_no_evil:

File: Makefile-docker [hunk 16]
```diff
@@ -162,7 +162,7 @@ setup-ui-tests:
 	# Create a proper superuser that can be used to access the API
 	python manage.py waffle_switch super-create-accounts on --create
 	python manage.py waffle_switch activate-autograph-signing on --create
-	python manage.py createsuperuser --email=uitest@restmail.net --username=uitest --noinput --add-to-supercreate-group --save-api-credentials=tests/ui/variables.json --hostname=olympia-frontend.test
+	python manage.py createsuperuser --email=uitest@restmail.net --username=uitest --noinput --add-to-supercreate-group --hostname=olympia-frontend.test
```
Comment (EnTeQuAk) [inline Makefile-docker]: This doesn't remove the functionality from the management command though. Can you remove that too please?

File: tox.ini (hunk 14 above)
Comment (jrbenny35) [reply tox.ini]: gah okay. I think atom saved old changes while I was rebasing.

File: tox.ini (hunk 15 above)
Comment (jrbenny35) [reply tox.ini]: 👀

File: Makefile-docker (hunk 16 above)
Comment (jrbenny35) [reply Makefile-docker]: We want to do that? I think it could still be useful to have.
